sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,333 @@
+
+    def transform(self, X):
+        """Transforms X to cluster-distance space.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape=(n_samples, n_features)
+            Data to transform.
+
+        Returns
+        -------
+        X_new : {array-like, sparse matrix}, shape=(n_samples, n_clusters)
+            X transformed in the new space.
+        """
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+
+        check_is_fitted(self, "cluster_centers_")
+
+        # Apply distance metric wrt. cluster centers (medoids),
+        # and return these distances
+        return pairwise_distances(X, Y=self.cluster_centers_,
+                                  metric=self.distance_metric)
+
+    def predict(self, X):
+        """Predict the closest cluster each sample in X belongs to.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            New data to predict.
+
+        Returns
+        -------
+        labels : array, shape [n_samples,]
+            Index of the cluster each sample belongs to.
+        """
+
+        check_is_fitted(self, "cluster_centers_")
+
+        # Check that the array is good and attempt to convert it to
+        # Numpy array if possible
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+
+        # Apply distance metric wrt. cluster centers (medoids)
+        distances = pairwise_distances(X, Y=self.cluster_centers_,
+                                       metric=self.distance_metric)
+
+        # Assign data points to clusters based on
+        # which cluster assignment yields
+        # the smallest distance
+        labels = np.argmin(distances, axis=1)
+
+        return labels
+
+    def _compute_inertia(self, X):
+        """Compute inertia of new samples. Inertia is defined as the sum of the
+        sample distances to closest cluster centers.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape=(n_samples, n_features)
+            Samples to compute inertia for.
+
+        Returns
+        -------
+        Sum of sample distances to closest cluster centers.
+
+        """
+
+        # Map the original X to the distance-space
+        Xt = self.transform(X)
+
+        # Define inertia as the sum of the sample-distances
+        # to closest cluster centers
+        inertia = np.sum(np.min(Xt, axis=1))
+
+        return inertia
+
+    def _get_initial_medoid_indices(self, distances, n_clusters):
+        """Initialize medoid indices using either a random or
+        heuristic approach.
+
+        """
+
+        if self.init == 'random':  # Random initialization
+
+            # Pick random k medoids as the initial ones.
+            medoids = self.random_state_.permutation(
+                distances.shape[0])[:n_clusters]
+
+        elif self.init == 'heuristic':  # Initialization by heuristic
+
```

Comment:
Could you avoid adding blank lines unless it would help separate two logical / functional blocks of code?